robot_smach_states/src/robot_smach_states/utility.py

```diff
@@ -351,6 +353,109 @@ def execute(self, userdata=None):
         return "done"
 
 
+class ResolveArm(smach.State):
+    def __init__(self, arm_designator, state_machine):
+        """ Resolves, if possible, an arm for a state machine taking into account all the arm requirements
+
+        :param arm_designator: given arm designator
+        :param state_machine: used state machine
+        """
+        smach.State.__init__(self,
+                             outcomes=['succeeded', 'failed'],
+                             output_keys=["arm"]
+                             )
+        self.arm_designator = arm_designator
+        self.state_machine = state_machine
+
+    def execute(self, userdata):
+        arm_requirements = collect_arm_requirements(self.state_machine)
+        resolved_arm = const_resolve(self.arm_designator, arm_requirements)
+        if resolved_arm is None:
+            rospy.logerror("Didn't find an arm")  # ToDo: improve error message
+            return "failed"
+        else:
+            userdata.arm = resolved_arm
+            return "succeeded"
+
+
+def check_arm_requirements(state_machine, robot):
+    """
+    Checks if the robot has an arm that meets the requirements of all children states of this state machine
+
+    :param state_machine: The state machine for which the requirements have to be checked
+    :param robot: Robot to use
+    :return: Check whether an arm is available that satisfies the requirements of the state machine
+    """
+    arm_requirements = collect_arm_requirements(state_machine)
+    try:
+        assert robot.get_arm(**arm_requirements) is not None,\
+            "None of the available arms meets all this state machine's requirements: {}".format(arm_requirements)
+    except AssertionError as e:
+        rospy.logerr("Getting arm requirements failed, arm requirements: {}".format(arm_requirements))
+        raise
+
+
+def collect_arm_requirements(state_machine):
+    """ Collects all requirements on the arm of this specific state machine
+
+    :param state_machine: State (machine) for which the requirements need to be collected
+    :return: All arm requirements of the state machine
+    """
+    def update_requirements(state):
+        """ Checks the input state for arm requirements and updates the current arm requirements if necessary
+
+        :param state: Smach state of which arm requirements should be checked against the current arm requirements
+        :return: current arm requirements, updated (if necessary) given the input state
+        """
+        for k, v in state.iteritems():
+            if k not in arm_requirements:
+                arm_requirements[k] = v
+            else:
+                for value in v:
+                    if value not in arm_requirements[k]:
+                        arm_requirements[k] += v
```

Comment:
This is what I meant when I proposed to split some of the functionality of const_resolve. Because this loop is very similar to the one below. 
Having said that. This is a readable code nitpick and not a critisism of your arms requirements approach.

Reply:
From what I see these two pieces of code are not mergeable.